From b951a1d64792de0d09057e1c2c566d37a4580ec1 Mon Sep 17 00:00:00 2001
From: Luca Chiodini <luca@chiodini.org>
Date: Tue, 27 Sep 2016 21:33:13 +0200
Subject: [PATCH] Add wiringPiI2CWriteBlock function

---
 wiringPi/wiringPiI2C.c | 14 ++++++++++++++
 wiringPi/wiringPiI2C.h |  2 ++
 2 files changed, 16 insertions(+)

diff --git a/wiringPi/wiringPiI2C.c b/wiringPi/wiringPiI2C.c
index c787bce..9c34212 100644
--- a/wiringPi/wiringPiI2C.c
+++ b/wiringPi/wiringPiI2C.c
@@ -230,3 +230,17 @@ int wiringPiI2CSetup (const int devId)
 
   return wiringPiI2CSetupInterface (device, devId) ;
 }
+
+int wiringPiI2CWriteBlock (int fd, int command, int data[], int n)
+{
+  int i = 0;
+  size_t quantity = n;
+  union i2c_smbus_data block;
+  // Limit data size.
+  if (I2C_SMBUS_I2C_BLOCK_MAX < quantity)
+    quantity = I2C_SMBUS_I2C_BLOCK_MAX;
+  block.block[0] = quantity;
+  for (; i < quantity; i++)
+    block.block[i+1] = data[i];
+  return i2c_smbus_access (fd, I2C_SMBUS_WRITE, command, I2C_SMBUS_I2C_BLOCK_DATA, &block) ;
+}
diff --git a/wiringPi/wiringPiI2C.h b/wiringPi/wiringPiI2C.h
index 6db8c68..c6b71d9 100644
--- a/wiringPi/wiringPiI2C.h
+++ b/wiringPi/wiringPiI2C.h
@@ -37,6 +37,8 @@ extern int wiringPiI2CWriteReg16     (int fd, int reg, int data) ;
 extern int wiringPiI2CSetupInterface (const char *device, int devId) ;
 extern int wiringPiI2CSetup          (const int devId) ;
 
+extern int wiringPiI2CWriteBlock     (int fd, int command, int data[], int n) ;
+
 #ifdef __cplusplus
 }
 #endif
-- 
2.7.4

